### ISSUE TITLE
ICE when constructing tuple

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#24892]
|  |  |
| --- | --- |
| Bugzilla Link | [24518](https://llvm.org/bz24518) |
| Version | 3.7 |
| OS | MacOS X |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
I tried compiling some code involving `std::tuple` using clang 3.7 (trunk 241984, downloaded via Macports) and clang 3.6.1 (also downloaded via Macports). Compiling under clang-3.7 results in an ICE. On the other hand, clang-3.6.1 produces a spurious error message. I've attached the preprocessed source file and run script resulting from the ICE.

I've looked into the spurious error message given by clang-3.6, but I have not been able to produce a MWE. In short, here's what's happening. I'm writing a multidimensional array library, and have a class like this:
```cpp
    template <class Func, class... Ts>
    struct elemwise_expr final
    {
    	std::tuple<const Ts&...> m_refs;
	const Func& m_func;
public:
	CC_ALWAYS_INLINE
	explicit elemwise_expr(const Func& f, const Ts&... ts)
	noexcept : m_refs{ts...}, m_func{f} {}

        // Code omitted.
    };
```
When I compile this code under clang-3.6.1, the compiler generates several `no matching constructor for initialization of ...` errors involving types from my library. After looking over my own code carefully, I realized that the error was generated by the `__is_constructible` test for the `tuple(_Up&&... __u)` constructor. It turns out that for some reason, this constructor appears to have been supplied only one argument -- a tuple -- despite the fact that none of the types `ts...` in my own code are tuples. Somehow, the arguments used to initialize `m_refs` are being redundantly nested in a tuple. The `__is_constructible` test then tries to invoke the constructors of various types in my library with a single tuple as an argument, and this fails. My guess is that this strange error becomes an ICE under clang 3.7.

Do you have any recommendations for workarounds, or how I could produce an MWE? I believe you should be able to compile the preprocessed source file under clang-3.6.1 instead of clang-3.7 to get the same results.

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [24518](https://llvm.org/bz24518) |
| Version | 3.7 |
| OS | MacOS X |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor |

## Extended Description 
I tried compiling some code involving `std::tuple` using clang 3.7 (trunk 241984, downloaded via Macports) and clang 3.6.1 (also downloaded via Macports). Compiling under clang-3.7 results in an ICE. On the other hand, clang-3.6.1 produces a spurious error message. I've attached the preprocessed source file and run script resulting from the ICE.

I've looked into the spurious error message given by clang-3.6, but I have not been able to produce a MWE. In short, here's what's happening. I'm writing a multidimensional array library, and have a class like this:
```cpp
    template &lt;class Func, class... Ts&gt;
    struct elemwise_expr final
    {
    	std::tuple&lt;const Ts&amp;...&gt; m_refs;
	const Func&amp; m_func;
public:
	CC_ALWAYS_INLINE
	explicit elemwise_expr(const Func&amp; f, const Ts&amp;... ts)
	noexcept : m_refs{ts...}, m_func{f} {}

        // Code omitted.
    };
```
When I compile this code under clang-3.6.1, the compiler generates several `no matching constructor for initialization of ...` errors involving types from my library. After looking over my own code carefully, I realized that the error was generated by the `__is_constructible` test for the `tuple(_Up&amp;&amp;... __u)` constructor. It turns out that for some reason, this constructor appears to have been supplied only one argument -- a tuple -- despite the fact that none of the types `ts...` in my own code are tuples. Somehow, the arguments used to initialize `m_refs` are being redundantly nested in a tuple. The `__is_constructible` test then tries to invoke the constructors of various types in my library with a single tuple as an argument, and this fails. My guess is that this strange error becomes an ICE under clang 3.7.

Do you have any recommendations for workarounds, or how I could produce an MWE? I believe you should be able to compile the preprocessed source file under clang-3.6.1 instead of clang-3.7 to get the same results.
</details>

**Comment by AaronBallman** [comment]
Closing the issue as incomplete due to the lack of a way to reproduce, but if the issue still happens on trunk, please reopen with a reproducer.
